### PR TITLE
[MOD-13920] Add find (range query) algorithm for numeric_range_tree

### PR DESCRIPTION
**Pull request by LukeMathWalker**
## Describe the changes in the pull request

Find implementation for `NumericRangeTree`.

Stacked on top #8276.

#### Which additional issues this PR fixes
1. MOD-...
2. #...

#### Main objects this PR modified
1. ...

#### Mark if applicable

- [ ] This PR introduces API changes
- [ ] This PR introduces serialization changes

#### Release Notes

- [ ] This PR requires release notes
- [x] This PR does not require release notes

If a release note is required (bug fix / new feature / enhancement), describe the **user impact** of this PR in the title.  

---

> [!NOTE]
> **Medium Risk**
> Adds a new public read-path (`NumericRangeTree::find`) whose traversal/pruning and pagination semantics directly affect query correctness and performance. Includes new overlap/containment helpers and invariant checks, but regressions could still surface under edge-case filters or large trees.
> 
> **Overview**
> Implements the numeric range tree read path by adding `NumericRangeTree::find`, which traverses the tree using **containment/overlap pruning** and returns ranges in ascending/descending order, with support for `NumericFilter` pagination via `offset`/`limit`.
> 
> Adds `NumericRange::contained_in`/`overlaps` helpers and (under `unittest`) new `find`-specific invariant checks (overlap, ordering/non-overlap, and limit sufficiency). Exposes `NumericFilter` from the crate and adds integration + property tests covering large-tree queries, empty/no-overlap cases, ordering, and pagination behavior.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 6bb03cff3f2ac8d118bd6cf1e120189461ded13a. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: src/redisearch_rs/numeric_range_tree/src/tree/find.rs [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+            }
+            NumericRangeNode::Leaf(leaf) => {
+                // Leaf node with overlap
+                if leaf.range.overlaps(min, max) {
+                    let num_docs = leaf.range.num_docs();
+                    *total += if *total == 0 && filter.offset == 0 {
+                        1
+                    } else {
+                        num_docs as usize
+                    };
```
[the remark this answers is not in the document]
Comment (LukeMathWalker) [reply src/redisearch_rs/numeric_range_tree/src/tree/find.rs]: This comes straight from the original C code, but I'm unsure as to why it is the way it is: https://github.com/RediSearch/RediSearch/blob/c20a8075afd7b74c45738f3b4e3400186058821b/src/numeric_index.c#L298
Any clue? @meiravgri

File: src/redisearch_rs/numeric_range_tree/src/tree/find.rs [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+        }
+
+        let node = &nodes[node_idx];
+        let min = filter.min;
+        let max = filter.max;
+
+        if let Some(range) = node.range() {
+            let num_docs = range.num_docs();
+            let contained = range.contained_in(min, max);
+            let overlaps = range.overlaps(min, max);
```
[the remark this answers is not in the document]
Comment (LukeMathWalker) [reply src/redisearch_rs/numeric_range_tree/src/tree/find.rs:94]: The definition of `NumericFilter` does include booleans for inclusiveness (https://github.com/RediSearch/RediSearch/blob/master/src/redisearch_rs/inverted_index/src/lib.rs#L87), but they are ignored in the existing C implementation for `numeric_index.c`. I _assume_ they are applied later in the iterator, since it's unlikely that the range boundaries are exactly the boundaries of our filter.
We could clarify `numeric_range_tree` by using a narrowed down filter type as input, thus removing ambiguity.

File: src/redisearch_rs/numeric_range_tree/src/tree/find.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (LukeMathWalker) [reply src/redisearch_rs/numeric_range_tree/src/tree/find.rs]: Thinking about this, I think I understand the rationale. I have documented it as a comment—please double-check it.

File: src/redisearch_rs/numeric_range_tree/src/tree/find.rs [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+                            internal.left_index(),
+                            filter,
+                            total,
+                        );
+                    }
+                }
+            }
+            NumericRangeNode::Leaf(leaf) => {
+                let num_docs = leaf.range.num_docs();
+                *total += if *total == 0 && filter.offset == 0 {
```
Comment (meiravgri) [inline src/redisearch_rs/numeric_range_tree/src/tree/find.rs:157]: Why is the += 1 special case only applied when offset == 0? If the concern is overcounting for partially-overlapping ranges, wouldn't the same issue apply when offset > 0?

File: src/redisearch_rs/numeric_range_tree/src/tree/find.rs [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+    /// retain their ranges up to `max_depth_range`.
+    ///
+    /// # Offset Handling
+    ///
+    /// The `filter.offset` and `filter.limit` parameters enable pagination.
+    /// We track the running total of documents and skip ranges until we've
+    /// passed the offset. See `recursive_find_ranges` for the special handling
+    /// of the first overlapping leaf.
+    pub fn find<'a>(&'a self, filter: &NumericFilter) -> Vec<&'a NumericRange> {
+        let mut ranges = Vec::with_capacity(8);
```
Comment (meiravgri) [inline src/redisearch_rs/numeric_range_tree/src/tree/find.rs:44]: any special reason for allocating 8 other then aligning with the c code?

File: src/redisearch_rs/numeric_range_tree/tests/integration/range.rs [hunk 5]
```diff
@@ -53,6 +53,39 @@ fn test_add_updates_cardinality() {
     );
 }
 
+#[test]
```
Comment (meiravgri) [inline src/redisearch_rs/numeric_range_tree/tests/integration/range.rs:56]: General comment - we should add tests (or modify existing) for a tree negative values and fp that are not exact integers

File: src/redisearch_rs/numeric_range_tree/src/tree/find.rs (hunk 4 above)
Comment (LukeMathWalker) [reply src/redisearch_rs/numeric_range_tree/src/tree/find.rs:44]: No special reason other than matching https://github.com/RediSearch/RediSearch/blob/2aad4f55511f4c8648fd918f5d556091ba16bcb3/src/numeric_index.c#L310.

File: src/redisearch_rs/numeric_range_tree/tests/integration/range.rs (hunk 5 above)
Comment (LukeMathWalker) [reply src/redisearch_rs/numeric_range_tree/tests/integration/range.rs:56]: All the property-based tests in https://github.com/RediSearch/RediSearch/blob/master/src/redisearch_rs/numeric_range_tree/tests/integration/properties.rs do use random floating point values, spanning from negative to positive, so that should give us coverage in that sense.

File: src/redisearch_rs/numeric_range_tree/src/tree/find.rs (hunk 3 above)
Comment (LukeMathWalker) [reply src/redisearch_rs/numeric_range_tree/src/tree/find.rs:157]: Let's keep it as it is for now, aligning with the C version. We'll revise in the contex of the optimizer iterator rework (https://redislabs.atlassian.net/browse/MOD-13983).
